File: src/front-end/src/infrastructure/transport/data-channel-framing.ts
// Envelope layout (little-endian, 17 bytes of overhead):
//   0-3  magic        "CKDC" (0x43 0x4B 0x44 0x43)
//   4    version      uint8, currently 1
//   5-8  message_id   uint32 LE, per-direction monotonic counter
//   9-10 chunk_index  uint16 LE, 0-based
//  11-12 chunk_count  uint16 LE, total chunks for this message_id
//  13-16 payload_len  uint32 LE, bytes in THIS chunk's payload
//  17+   payload      N bytes

const MAGIC_0 = 0x43;
const MAGIC_1 = 0x4b;
const MAGIC_2 = 0x44;
const MAGIC_3 = 0x43;

const VERSION = 1;
const HEADER_SIZE = 17;

export const CHUNK_PAYLOAD_MAX = 65519; // 65536 - HEADER_SIZE
const MAX_MESSAGE_BYTES = 16 * 1024 * 1024;
const MAX_CHUNKS = Math.ceil(MAX_MESSAGE_BYTES / CHUNK_PAYLOAD_MAX); // 257
const CHUNK_ASSEMBLY_TIMEOUT_MS = 5_000;
const MAX_IN_FLIGHT_MESSAGES = 16;

let _nextMessageId = 0;

export function nextMessageId(): number {
  _nextMessageId = (_nextMessageId + 1) >>> 0; // uint32 wrap-around
  return _nextMessageId;
}

export function packMessage(messageId: number, payload: Uint8Array): ArrayBuffer[] {
  if (payload.byteLength > MAX_MESSAGE_BYTES) {
    throw new Error(
      `Message too large: ${payload.byteLength} bytes exceeds limit of ${MAX_MESSAGE_BYTES}`
    );
  }

  const chunkCount = payload.byteLength === 0 ? 1 : Math.ceil(payload.byteLength / CHUNK_PAYLOAD_MAX);
  const chunks: ArrayBuffer[] = [];

  for (let i = 0; i < chunkCount; i++) {
    const start = i * CHUNK_PAYLOAD_MAX;
    const slice = payload.subarray(start, start + CHUNK_PAYLOAD_MAX);
    const buf = new ArrayBuffer(HEADER_SIZE + slice.byteLength);
    const view = new DataView(buf);
    const bytes = new Uint8Array(buf);

    bytes[0] = MAGIC_0;
    bytes[1] = MAGIC_1;
    bytes[2] = MAGIC_2;
    bytes[3] = MAGIC_3;
    view.setUint8(4, VERSION);
    view.setUint32(5, messageId, /* littleEndian= */ true);
    view.setUint16(9, i, true);
    view.setUint16(11, chunkCount, true);
    view.setUint32(13, slice.byteLength, true);
    bytes.set(slice, HEADER_SIZE);

    chunks.push(buf);
  }

  return chunks;
}

interface InFlightMessage {
  chunks: (Uint8Array | null)[];
  chunkCount: number;
  received: number;
  timer: ReturnType<typeof setTimeout>;
}

export class ChunkReassembler {
  private readonly inFlight = new Map<number, InFlightMessage>();
  private readonly inFlightOrder: number[] = []; // oldest first

  pushChunk(buffer: ArrayBuffer): Uint8Array | null {
    if (buffer.byteLength < HEADER_SIZE) {
      console.warn('[framing] Chunk too small, dropping');
      return null;
    }

    const view = new DataView(buffer);
    const bytes = new Uint8Array(buffer);

    if (bytes[0] !== MAGIC_0 || bytes[1] !== MAGIC_1 || bytes[2] !== MAGIC_2 || bytes[3] !== MAGIC_3) {
      console.warn('[framing] Bad magic bytes, dropping chunk');
      return null;
    }

    const version = view.getUint8(4);
    if (version !== VERSION) {
      console.warn(`[framing] Unknown version ${version}, dropping chunk`);
      return null;
    }

    const messageId = view.getUint32(5, true);
    const chunkIndex = view.getUint16(9, true);
    const chunkCount = view.getUint16(11, true);
    const payloadLen = view.getUint32(13, true);

    if (chunkCount < 1 || chunkCount > MAX_CHUNKS) {
      console.warn(`[framing] Invalid chunk_count ${chunkCount}, dropping`);
      return null;
    }

    if (HEADER_SIZE + payloadLen > buffer.byteLength) {
      console.warn(`[framing] payload_len ${payloadLen} exceeds buffer size, dropping`);
      return null;
    }

    if (chunkIndex >= chunkCount) {
      console.warn(`[framing] chunk_index ${chunkIndex} >= chunk_count ${chunkCount}, dropping`);
      return null;
    }

    let msg = this.inFlight.get(messageId);

    if (!msg) {
      if (this.inFlightOrder.length >= MAX_IN_FLIGHT_MESSAGES) {
        const oldest = this.inFlightOrder.shift()!;
        const evicted = this.inFlight.get(oldest);
        if (evicted) {
          clearTimeout(evicted.timer);
          this.inFlight.delete(oldest);
          console.warn(`[framing] Evicted oldest in-flight message_id=${oldest} (cap reached)`);
        }
      }

      msg = {
        chunks: new Array<Uint8Array | null>(chunkCount).fill(null),
        chunkCount,
        received: 0,
        timer: setTimeout(() => {
          const m = this.inFlight.get(messageId);
          if (m) {
            this.inFlight.delete(messageId);
            const idx = this.inFlightOrder.indexOf(messageId);
            if (idx >= 0) {
              this.inFlightOrder.splice(idx, 1);
            }
            console.warn(
              `[framing] Timeout for message_id=${messageId}: received ${m.received}/${m.chunkCount} chunks`
            );
          }
        }, CHUNK_ASSEMBLY_TIMEOUT_MS),
      };
      this.inFlight.set(messageId, msg);
      this.inFlightOrder.push(messageId);
    }

    if (msg.chunkCount !== chunkCount) {
      console.warn(`[framing] chunk_count mismatch for message_id=${messageId}, dropping`);
      return null;
    }

    if (msg.chunks[chunkIndex] !== null) {
      return null; // duplicate — ignore silently
    }

    msg.chunks[chunkIndex] = new Uint8Array(buffer, HEADER_SIZE, payloadLen).slice();
    msg.received++;

    if (msg.received < msg.chunkCount) {
      return null;
    }

    clearTimeout(msg.timer);
    this.inFlight.delete(messageId);
    const orderIdx = this.inFlightOrder.indexOf(messageId);
    if (orderIdx >= 0) {
      this.inFlightOrder.splice(orderIdx, 1);
    }

    let total = 0;
    for (const c of msg.chunks) {
      total += c!.byteLength;
    }
    const result = new Uint8Array(total);
    let offset = 0;
    for (const c of msg.chunks) {
      result.set(c!, offset);
      offset += c!.byteLength;
    }
    return result;
  }
}

File: src/cpp_accelerator/adapters/webrtc/data_channel_framing.cpp
#include "src/cpp_accelerator/adapters/webrtc/data_channel_framing.h"

#include <algorithm>
#include <cstring>
#include <stdexcept>
#include <string>

#include <spdlog/spdlog.h>

namespace jrb::adapters::webrtc {

namespace {

// Envelope layout (little-endian, 17 bytes):
//   0-3   magic       "CKDC" = {0x43, 0x4B, 0x44, 0x43}
//   4     version     uint8, currently 1
//   5-8   message_id  uint32 LE
//   9-10  chunk_index uint16 LE, 0-based
//  11-12  chunk_count uint16 LE
//  13-16  payload_len uint32 LE, bytes in THIS chunk's payload
//  17+    payload

constexpr uint8_t kMagic[4] = {0x43, 0x4B, 0x44, 0x43};
constexpr uint8_t kVersion = 1;
constexpr size_t kHeaderSize = 17;

void WriteU32LE(std::byte* dst, uint32_t v) {
  dst[0] = static_cast<std::byte>(v & 0xffu);
  dst[1] = static_cast<std::byte>((v >> 8u) & 0xffu);
  dst[2] = static_cast<std::byte>((v >> 16u) & 0xffu);
  dst[3] = static_cast<std::byte>((v >> 24u) & 0xffu);
}

void WriteU16LE(std::byte* dst, uint16_t v) {
  dst[0] = static_cast<std::byte>(v & 0xffu);
  dst[1] = static_cast<std::byte>((v >> 8u) & 0xffu);
}

uint32_t ReadU32LE(const std::byte* src) {
  return static_cast<uint32_t>(src[0]) |
         (static_cast<uint32_t>(src[1]) << 8u) |
         (static_cast<uint32_t>(src[2]) << 16u) |
         (static_cast<uint32_t>(src[3]) << 24u);
}

uint16_t ReadU16LE(const std::byte* src) {
  return static_cast<uint16_t>(static_cast<uint16_t>(src[0]) |
                                (static_cast<uint16_t>(src[1]) << 8u));
}

}  // namespace

std::vector<std::vector<std::byte>> PackMessage(
    uint32_t message_id, std::span<const std::byte> payload) {
  if (payload.size() > kMaxMessageBytes) {
    throw std::invalid_argument(
        "PackMessage: payload too large (" + std::to_string(payload.size()) +
        " bytes), max is " + std::to_string(kMaxMessageBytes));
  }

  const uint32_t total = static_cast<uint32_t>(payload.size());
  const uint16_t chunk_count =
      total == 0u
          ? 1u
          : static_cast<uint16_t>((total + kChunkPayloadMax - 1u) / kChunkPayloadMax);

  std::vector<std::vector<std::byte>> chunks;
  chunks.reserve(chunk_count);

  for (uint16_t i = 0; i < chunk_count; ++i) {
    const uint32_t start = static_cast<uint32_t>(i) * kChunkPayloadMax;
    const uint32_t len =
        (total == 0u) ? 0u : std::min(kChunkPayloadMax, total - start);

    std::vector<std::byte> chunk(kHeaderSize + len);

    std::memcpy(chunk.data(), kMagic, 4);
    chunk[4] = static_cast<std::byte>(kVersion);
    WriteU32LE(chunk.data() + 5, message_id);
    WriteU16LE(chunk.data() + 9, i);
    WriteU16LE(chunk.data() + 11, chunk_count);
    WriteU32LE(chunk.data() + 13, len);
    if (len > 0u) {
      std::memcpy(chunk.data() + kHeaderSize, payload.data() + start, len);
    }

    chunks.push_back(std::move(chunk));
  }

  return chunks;
}

ChunkReassembler::ChunkReassembler(std::chrono::milliseconds timeout)
    : timeout_(timeout) {}

void ChunkReassembler::EvictTimedOut() {
  const auto now = std::chrono::steady_clock::now();
  std::vector<uint32_t> to_evict;

  for (const auto& [id, msg] : in_flight_) {
    if (now - msg.first_chunk_at > timeout_) {
      spdlog::warn("[framing] Timeout for message_id={}: received {}/{} chunks",
                   id, msg.received, msg.chunk_count);
      to_evict.push_back(id);
    }
  }

  for (const uint32_t id : to_evict) {
    in_flight_.erase(id);
    in_flight_order_.erase(
        std::remove(in_flight_order_.begin(), in_flight_order_.end(), id),
        in_flight_order_.end());
  }
}

void ChunkReassembler::EvictOldestIfFull() {
  if (static_cast<int>(in_flight_.size()) < kMaxInFlightMessages) {
    return;
  }
  const uint32_t oldest = in_flight_order_.front();
  in_flight_order_.erase(in_flight_order_.begin());
  in_flight_.erase(oldest);
  spdlog::warn("[framing] Evicted oldest in-flight message_id={} (cap={})",
               oldest, kMaxInFlightMessages);
}

std::optional<std::vector<std::byte>> ChunkReassembler::PushChunk(
    std::span<const std::byte> raw) {
  EvictTimedOut();

  if (raw.size() < kHeaderSize) {
    spdlog::warn("[framing] Chunk too small ({} bytes), dropping", raw.size());
    return std::nullopt;
  }

  if (std::memcmp(raw.data(), kMagic, 4) != 0) {
    spdlog::warn("[framing] Bad magic bytes, dropping chunk");
    return std::nullopt;
  }

  const uint8_t version = static_cast<uint8_t>(raw[4]);
  if (version != kVersion) {
    spdlog::warn("[framing] Unknown version {}, dropping chunk", version);
    return std::nullopt;
  }

  const uint32_t message_id = ReadU32LE(raw.data() + 5);
  const uint16_t chunk_index = ReadU16LE(raw.data() + 9);
  const uint16_t chunk_count = ReadU16LE(raw.data() + 11);
  const uint32_t payload_len = ReadU32LE(raw.data() + 13);

  if (chunk_count < 1u || chunk_count > kMaxChunks) {
    spdlog::warn("[framing] Invalid chunk_count={}, dropping", chunk_count);
    return std::nullopt;
  }

  if (kHeaderSize + payload_len > raw.size()) {
    spdlog::warn("[framing] payload_len={} exceeds buffer size={}, dropping",
                 payload_len, raw.size());
    return std::nullopt;
  }

  if (chunk_index >= chunk_count) {
    spdlog::warn("[framing] chunk_index={} >= chunk_count={}, dropping",
                 chunk_index, chunk_count);
    return std::nullopt;
  }

  auto it = in_flight_.find(message_id);
  if (it == in_flight_.end()) {
    EvictOldestIfFull();
    InFlightMessage msg;
    msg.chunk_count = chunk_count;
    msg.received = 0;
    msg.first_chunk_at = std::chrono::steady_clock::now();
    msg.chunks.resize(chunk_count);
    in_flight_.emplace(message_id, std::move(msg));
    in_flight_order_.push_back(message_id);
    it = in_flight_.find(message_id);
  }

  InFlightMessage& msg = it->second;

  if (msg.chunk_count != chunk_count) {
    spdlog::warn("[framing] chunk_count mismatch for message_id={}, dropping",
                 message_id);
    return std::nullopt;
  }

  if (msg.chunks[chunk_index].has_value()) {
    return std::nullopt;  // duplicate — ignore silently
  }

  const auto* payload_ptr = raw.data() + kHeaderSize;
  msg.chunks[chunk_index] =
      std::vector<std::byte>(payload_ptr, payload_ptr + payload_len);
  ++msg.received;

  if (msg.received < msg.chunk_count) {
    return std::nullopt;
  }

  // All chunks present — concatenate in order and return
  size_t total = 0;
  for (const auto& c : msg.chunks) {
    total += c->size();
  }
  std::vector<std::byte> result;
  result.reserve(total);
  for (const auto& c : msg.chunks) {
    result.insert(result.end(), c->begin(), c->end());
  }

  spdlog::debug("[framing] Message {} complete: {} bytes in {} chunks",
                message_id, total, chunk_count);

  in_flight_.erase(it);
  in_flight_order_.erase(
      std::remove(in_flight_order_.begin(), in_flight_order_.end(), message_id),
      in_flight_order_.end());

  return result;
}

}  // namespace jrb::adapters::webrtc
